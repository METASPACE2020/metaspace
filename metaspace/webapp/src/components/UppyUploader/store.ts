// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from 'vue'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State | any>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach((listener) => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
