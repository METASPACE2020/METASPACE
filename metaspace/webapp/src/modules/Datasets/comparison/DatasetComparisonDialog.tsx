import { computed, defineComponent, reactive } from '@vue/composition-api'
import { Workflow, WorkflowStep } from '../../../components/Workflow'
import { Select, Option, InputNumber, Button, Dialog } from '../../../lib/element-ui'
import { ErrorLabelText } from '../../../components/Form'
import { useQuery } from '@vue/apollo-composable'
import {
  DatasetDetailItem,
  datasetDetailItemsQuery,
} from '../../../api/dataset'
import './DatasetComparisonDialog.scss'
import { encodeParams } from '../../Filters'

interface DatasetComparisonDialogProps {
  title: string
  selectedDatasetIds: string[]
  firstStepError: string
  firstStepLabel: string
  secondStepLabel: string
  thirdStepLabel: string
  finalStepError: string
}

interface DatasetComparisonDialogState {
  selectedDatasetIds: string[]
  workflowStep: number
  nCols: number
  nRows: number
  showOptions: boolean
  firstStepError: boolean
  finalStepError: boolean
  arrangement: {[key: string] : string}
}

export const DatasetComparisonDialog = defineComponent<DatasetComparisonDialogProps>({
  name: 'DatasetComparisonDialog',
  props: {
    title: {
      type: String,
      default: 'Datasets Comparison',
    },
    firstStepError: {
      type: String,
      default: 'Please select at least two datasets to be compared!',
    },
    finalStepError: {
      type: String,
      default: 'Please place all the selected datasets on the grid!',
    },
    firstStepLabel: {
      type: String,
      default: 'Select the datasets',
    },
    secondStepLabel: {
      type: String,
      default: 'Set the grid arrangement',
    },
    thirdStepLabel: {
      type: String,
      default: 'Choose the datasets disposition',
    },
    selectedDatasetIds: {
      type: Array,
      default: () => [],
    },
  },
  setup(props, { emit, root }) {
    const state = reactive<DatasetComparisonDialogState>({
      selectedDatasetIds: props.selectedDatasetIds,
      workflowStep: 1,
      nCols: 2,
      nRows: 2,
      showOptions: true,
      firstStepError: false,
      finalStepError: false,
      arrangement: {},
    })
    const {
      result: datasetResult,
      loading: datasetLoading,
    } = useQuery<{allDatasets: DatasetDetailItem}>(datasetDetailItemsQuery, {
      dFilter: { ids: null, polarity: null, metadataType: 'Imaging MS' },
      query: '',
      inpFdrLvls: [10],
      checkLvl: 10,
      offset: 0,
      orderBy: 'ORDER_BY_DATE',
      sortingOrder: 'DESCENDING',
    })
    const dataset = computed(() => datasetResult.value != null ? datasetResult.value.allDatasets : null)

    const annotationsLink = (datasetIds: string[]) => {
      const query = {
        datasetIds: datasetIds,
      }

      return {
        name: 'datasets-comparison',
        query: encodeParams(query),
        params: {
          nCols: state.nCols.toString(),
          nRows: state.nRows.toString(),
          grid: JSON.stringify(state.arrangement),
        },
      }
    }

    const handleDatasetSelection = (options: string[]) => {
      state.selectedDatasetIds = options
    }

    const handleSelection = (value: string, row: number, col: number) => {
      state.showOptions = false
      state.arrangement[`${row}-${col}`] = value
      setTimeout(() => {
        state.showOptions = true
      }, 0)
    }

    return () => {
      return (
        <Dialog
          visible
          lockScroll={true}
          onclose={() => emit('close')}
          class="dataset-comparison-dialog sm-content-page el-dialog-lean">
          <h1>{props.title}</h1>
          <Workflow>
            <WorkflowStep
              active={state.workflowStep === 1}
              done={state.workflowStep > 1}
            >
              <p class="sm-workflow-header">
                {props.firstStepLabel}
              </p>
              {
                state.workflowStep === 1
                && <form>
                  <Select
                    class={`w-full ${state.firstStepError ? 'sm-form-error' : ''}`}
                    value={state.selectedDatasetIds}
                    multiple
                    placeholder=" "
                    onChange={handleDatasetSelection}>
                    {
                      Array.isArray(dataset?.value)
                        && dataset?.value.map((ds) => {
                          return (
                            <Option key={ds.id} label={ds.name} value={ds.id}/>
                          )
                        })
                    }
                  </Select>
                  {
                    state.firstStepError
                    && <ErrorLabelText class='mt-0'>{props.firstStepError}</ErrorLabelText>
                  }
                  <Button onClick={() => {
                    if (state.selectedDatasetIds.length > 1) {
                      state.firstStepError = false
                      state.workflowStep = 2
                    } else {
                      state.firstStepError = true
                    }
                  }} type="primary">
                    Next
                  </Button>
                </form>
              }
            </WorkflowStep>
            <WorkflowStep
              active={state.workflowStep === 2}
              done={state.workflowStep > 2}
            >
              <p class="sm-workflow-header">
                {props.secondStepLabel}

              </p>
              {
                state.workflowStep === 2
                  && <form>
                    <div class='w-full flex flex-row items-center justify-center'>
                      <div class='m-2'>
                        <p>Number of rows</p>
                        <InputNumber
                          size="mini"
                          min={1}
                          max={state.nCols * state.nRows >= 15 ? state.nRows : undefined}
                          value={state.nRows}
                          onChange={(value: number) => { state.nRows = value }}/>
                      </div>
                      <div class='m-2'>
                        <p>Number of columns</p>
                        <InputNumber
                          size="mini"
                          min={1}
                          max={state.nCols * state.nRows >= 15 ? state.nCols : undefined}
                          value={state.nCols}
                          onChange={(value: number) => { state.nCols = value }}/>
                      </div>
                    </div>
                    <div class='dataset-comparison-dialog-grid'>
                      {Array.from(Array(state.nRows).keys()).map((row) => {
                        return (
                          <div key={row} class='dataset-comparison-dialog-row'>
                            {Array.from(Array(state.nCols).keys()).map((col) => {
                              return (
                                <div key={col} class='dataset-comparison-dialog-col'>
                                  {(col + 1) + (row * state.nCols)}
                                </div>
                              )
                            })}
                          </div>
                        )
                      })}
                    </div>
                    <Button onClick={() => { state.workflowStep = 1 }}>
                      Prev
                    </Button>
                    <Button onClick={() => { state.workflowStep = 3 }} type="primary">
                      Next
                    </Button>
                  </form>
              }
            </WorkflowStep>

            <WorkflowStep
              active={state.workflowStep === 3}
              done={state.workflowStep > 3}
            >
              <p class="sm-workflow-header">
                {props.thirdStepLabel}

              </p>
              {
                state.workflowStep === 3
                && <form>
                  <div class='dataset-comparison-dialog-grid'>
                    {Array.from(Array(state.nRows).keys()).map((row) => {
                      return (
                        <div key={row} class='dataset-comparison-dialog-row'>
                          {Array.from(Array(state.nCols).keys()).map((col) => {
                            return (
                              <div key={col} class='dataset-comparison-dialog-col'>
                                <Select
                                  class={`dataset-cell ${state.finalStepError ? 'sm-form-error' : ''}`}
                                  value={state.arrangement[`${row}-${col}`]}
                                  placeholder=" "
                                  clearable
                                  onChange={(value: string) => { handleSelection(value, row, col) }}>
                                  {
                                    state.showOptions
                                    && Array.isArray(dataset?.value)
                                    && dataset?.value.map((ds) => {
                                      return (
                                        <Option
                                          class='dataset-cell-option'
                                          disabled={Object.values(state.arrangement).includes(ds.id)}
                                          key={ds.id} label={ds.name} value={ds.id}/>
                                      )
                                    })
                                  }
                                </Select>
                              </div>
                            )
                          })}
                        </div>
                      )
                    })}
                  </div>
                  {
                    state.finalStepError
                    && <ErrorLabelText class='mt-0'>{props.finalStepError}</ErrorLabelText>
                  }
                  <Button onClick={() => { state.workflowStep = 2 }}>
                    Prev
                  </Button>
                  <Button onClick={() => {
                    if (Object.values(state.arrangement).length < state.selectedDatasetIds.length) {
                      state.finalStepError = true
                    } else {
                      root.$router.push(annotationsLink(state.selectedDatasetIds.sort()))
                    }
                  }} type="primary">
                    Compare
                  </Button>
                </form>
              }
            </WorkflowStep>
          </Workflow>
        </Dialog>
      )
    }
  },
})
