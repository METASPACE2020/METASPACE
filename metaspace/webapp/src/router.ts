/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter, { RawLocation } from 'vue-router'
import AboutPage from './modules/App/AboutPage.vue'
import DatasetsPage from './modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from './modules/Account'
import { redirectAfterSignIn } from './modules/Account/signInReturnUrl'
import { updateDBParam } from './modules/Filters/url'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Annotations/AnnotationsPage.vue'),
  DatasetSummary: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ './modules/Datasets/summary/DatasetSummary.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ './modules/MetadataEditor/MetadataEditPage.vue'),
  ImageAlignmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ './modules/ImageAlignment/ImageAlignmentPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ './modules/MetadataEditor/UploadPage.vue'),
  DesignPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DesignPage" */ './design/DesignPage.vue'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/HelpPage.vue'),
  EditUserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/GroupProfile/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/GroupsListPage.vue'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/TermsPage.vue'),
  PublicationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PublicationsPage.vue'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/GroupProfile/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Project/ViewProjectPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(about)?', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
    { path: '/annotations', component: asyncPages.AnnotationsPage },
    {
      path: '/datasets',
      component: DatasetsPage,
      children: [
        { path: '', component: asyncPages.DatasetTable },
        { path: 'summary', component: asyncPages.DatasetSummary },
      ],
    },
    { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
    { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.ImageAlignmentPage },
    { path: '/upload', component: asyncPages.UploadPage },
    { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },
    { path: '/user/me', component: asyncPages.EditUserPage },

    { path: '/admin/health', component: asyncPages.SystemHealthPage },
    { path: '/admin/groups', component: asyncPages.GroupsListPage },

    { path: '/account/sign-in', component: DialogPage, props: { dialog: 'signIn' } },
    { path: '/account/sign-in-success', redirect: redirectAfterSignIn },
    { path: '/account/create-account', component: DialogPage, props: { dialog: 'createAccount' } },
    { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
    { path: '/account/reset-password', component: ResetPasswordPage },

    { path: '/group/create', component: asyncPages.CreateGroupPage },
    { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

    { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
    { // Legacy URL sent in "request access" emails up until Feb 2019
      path: '/project/:projectIdOrSlug/manage',
      redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } } as RawLocation,
    },
    { path: '/projects', component: asyncPages.ProjectsListPage },

    { path: '/terms', component: asyncPages.TermsPage, meta: { footer: true } },
    { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
    { path: '/publications', component: asyncPages.PublicationsPage, meta: { footer: true } },

    { path: '/design', component: asyncPages.DesignPage, meta: { footer: true, flex: true } },

    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
