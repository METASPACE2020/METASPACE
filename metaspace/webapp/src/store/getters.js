import {mzFilterPrecision} from '../util';
import {decodeParams, decodeSettings, getLevel} from '../modules/Filters';
import config from '../config';

/** For filters where empty string is a valid client-side value that must be converted to empty string for the API */
const noneToEmptyString = s => s === 'none' ? '' : s;

export default {
  filterLevel(state) {
    return getLevel(state.route.path);
  },

  filter(state) {
    return decodeParams(state.route, state.filterLists);
  },

  settings(state) {
    return decodeSettings(state.route);
  },

  ftsQuery(state, getters) {
    return getters.filter.simpleQuery;
  },

  gqlAnnotationFilter(state, getters) {
    const filter = getters.filter;
    const colocalizationAlgo = getters.settings.annotationView.colocalizationAlgo;

    const f = {
      database: filter.database,
      compoundQuery: filter.compoundName,
      chemMod: noneToEmptyString(filter.chemMod),
      neutralLoss: noneToEmptyString(filter.neutralLoss),
      adduct: filter.adduct,
      fdrLevel: filter.fdrLevel,
      colocalizedWith: filter.colocalizedWith,
      // Only include colocalizationAlgo if there is another filter that uses it. Otherwise the annotations list
      // refreshes unnecessarily when changing algorithm.
      colocalizationAlgo: filter.colocalizedWith || filter.colocalizationSamples ? colocalizationAlgo : null,
      colocalizationSamples: filter.colocalizationSamples,
      offSample: filter.offSample == null ? undefined : !!filter.offSample,
    };

    if (!config.features.neutral_losses && !config.features.advanced_ds_config) {
      f.hasNeutralLoss = false;
    }

    if (!config.features.chem_mods && !config.features.advanced_ds_config) {
      f.hasChemMod = false;
    }

    if (!config.features.all_adducts) {
      f.hasHiddenAdduct = false;
    }

    if (filter.minMSM)
      f.msmScoreFilter = {min: filter.minMSM, max: 1.0};

    if (filter.mz) {
      const mz = parseFloat(filter.mz),
            deltamz = parseFloat(mzFilterPrecision(mz));
      f.mzFilter = {
        min: mz - deltamz,
        max: mz + deltamz
      };
    }

    return f;
  },

  gqlDatasetFilter(state, getters) {
    const filter = getters.filter;
    const {group, project, submitter, datasetIds, polarity,
           organism, organismPart, condition, growthConditions,
           ionisationSource, analyzerType, maldiMatrix, metadataType,
           compoundName} = filter;
    const level = getters.filterLevel;
    const hasAnnotationMatching = level === 'dataset' && compoundName ? { compoundQuery: compoundName } : undefined;
    return {
      group: group,
      project: project,
      submitter: submitter,

      // temporary workaround because of array-related bugs in apollo-client
      ids: datasetIds ? datasetIds.join("|") : null,

      organism,
      organismPart,
      condition,
      growthConditions,
      ionisationSource,
      maldiMatrix,
      analyzerType,
      polarity: polarity ? polarity.toUpperCase() : null,
      metadataType,
      hasAnnotationMatching,
    }
  },

  gqlColocalizationFilter(state, getters) {
    const {datasetIds, colocalizedWith, database, fdrLevel} = getters.filter;
    const colocalizationAlgo = getters.settings.annotationView.colocalizationAlgo;
    if (datasetIds && !datasetIds.includes('|') && colocalizedWith != null && database != null && fdrLevel != null) {
      return {colocalizedWith, colocalizationAlgo, database, fdrLevel};
    } else {
      return null;
    }
  },
}
