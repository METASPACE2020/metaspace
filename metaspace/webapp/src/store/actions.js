import { fetchOptionListsQuery } from '../api/metadata'
import { decodeParams } from '../modules/Filters'
import { computed } from 'vue'
import store from '../store/index'

export default {
  async initFilterLists(context, apolloClient) {
    if (context.state.filterListsLoading || context.state.filterLists != null) return
    await context.commit('setFilterListsLoading')

    const response = await apolloClient.query({
      query: fetchOptionListsQuery,
      fetchPolicy: 'cache-first',
    })

    // set annotationIds default according to values passed from snapshot to the store state
    context.commit('setFilterLists', {
      ...response.data,
      annotationIds: computed(() => store.state.snapshotAnnotationIds),
    })

    // Refresh the current filter so that computed defaults that depend on `filterLists` are applied
    const filter = decodeParams(context.state.route, context.state.filterLists)

    context.commit('updateFilter', filter)
  },
}
