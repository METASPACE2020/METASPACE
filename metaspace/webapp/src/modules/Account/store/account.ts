import { Module } from 'vuex';
import { DialogType } from '../dialogs';
import router from '../../../router';

export interface AccountState {
  dialog: DialogType | null;
  showDialogAsPage: boolean;
}

const account = {
  namespaced: true,
  state: {
    dialog: null,
    showDialogAsPage: false
  },
  mutations: {
    showDialog(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = false;
    },
    showDialogAsPage(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = true;
    },
    hideDialog(state: AccountState, dialog: DialogType) {
      if (state.dialog === dialog) {
        state.dialog = null;
        if (state.showDialogAsPage) {
          router.push('/');
        }
      }
    }
  }
};

export default account as Module<AccountState, any>;
