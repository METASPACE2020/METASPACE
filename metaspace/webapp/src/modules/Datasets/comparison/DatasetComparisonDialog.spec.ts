import { mount } from '@vue/test-utils'
import router from '../../../router'
import store from '../../../store/index'
import Vue from 'vue'
import Vuex from 'vuex'
import { sync } from 'vuex-router-sync'
import { DatasetComparisonDialog } from './DatasetComparisonDialog'
import { initMockGraphqlClient, apolloProvider } from '../../../../tests/utils/mockGraphqlClient'

Vue.use(Vuex)
sync(store, router)

describe('DatasetComparisonDialog', () => {
  const propsData = {
    selectedDatasetIds: [
      '2021-04-06_08h35m04s',
    ],
  }

  const graphqlReturnData = [{
    id: '2021-03-31_11h02m28s',
    name: 'New 3 (1)',
    uploadDT: '2021-03-31T14:02:28.722Z',
  }, {
    id: '2021-03-30_18h25m18s',
    name: 'Untreated_3_434_super_lite_19_31 (1)',
    uploadDT: '2021-03-30T21:25:18.473Z',
  }, {
    id: '2021-04-06_08h35m04s',
    name: 'Untreated_3_434_super_lite_19_31 (1)',
    uploadDT: '2021-03-30T21:25:18.473Z',
  }]

  const testHarness = Vue.extend({
    components: {
      DatasetComparisonDialog,
    },
    render(h) {
      return h(DatasetComparisonDialog, { props: this.$attrs })
    },
  })
  const graphqlWithData = () => {
    initMockGraphqlClient({
      Query: () => ({
        allDatasets: () => {
          return graphqlReturnData
        },
      }),
    })
  }

  it('it should match snapshot', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  it('it should display the minimum required datasets error', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()
    wrapper.find('.el-button--primary').trigger('click')
    await Vue.nextTick()

    expect(wrapper.find('.text-danger').exists()).toBe(true)
    expect(wrapper.find('.text-danger').text())
      .toBe('Please select at least two datasets to be compared!')
  })

  it('it should select a dataset and advance to the next step', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    wrapper.find('.el-select').trigger('click')
    wrapper.find('.el-select-dropdown__item').trigger('click')
    wrapper.find('.el-button--primary').trigger('click')
    await Vue.nextTick()

    expect(wrapper.find('.text-danger').exists()).toBe(false)
    expect(wrapper.findAll('.sm-workflow-step').at(1).classes()
      .includes('active')).toBe(true)
  })

  it('it should check the ds comparison grid nofcols and rows', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    // advances first step
    wrapper.find('.el-select').trigger('click')
    wrapper.find('.el-select-dropdown__item').trigger('click')
    wrapper.find('.el-button--primary').trigger('click')
    await Vue.nextTick()

    expect(wrapper.findAll('.dataset-comparison-dialog-row').length).toBe(2)
    expect(wrapper.findAll('.dataset-comparison-dialog-col').length).toBe(4)
  })

  it('it should display the final step required error', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    // advances first and second step and tries to compare
    wrapper.find('.el-select').trigger('click')
    wrapper.find('.el-select-dropdown__item').trigger('click')
    wrapper.find('.el-button--primary').trigger('click')
    await Vue.nextTick()
    wrapper.find('.el-button--primary').trigger('click')
    await Vue.nextTick()
    wrapper.find('.el-button--primary').trigger('click')
    await Vue.nextTick()

    expect(wrapper.find('.text-danger').exists()).toBe(true)
    expect(wrapper.find('.text-danger').text())
      .toBe('Please place all the selected datasets on the grid!')
  })

  it('it should set ds placement on the grid and advance to comparison page', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, propsData })
    await Vue.nextTick()

    // advances first and second step
    wrapper.find('.el-select').trigger('click')
    wrapper.find('.el-select-dropdown__item').trigger('click')
    await Vue.nextTick()
    wrapper.find('.el-select').trigger('click')
    wrapper.findAll('.el-select-dropdown__item').at(1).trigger('click')
    wrapper.find('.el-button--primary').trigger('click')
    await Vue.nextTick()
    wrapper.find('.el-button--primary').trigger('click')
    await Vue.nextTick()

    // place datasets
    wrapper.findAll('.dataset-cell').at(0).trigger('click')
    wrapper.findAll('.el-select-dropdown__item').at(0).trigger('click')
    await Vue.nextTick()
    await new Promise((resolve) => setTimeout(resolve, 500))

    wrapper.findAll('.dataset-cell').at(1).trigger('click')
    wrapper.findAll('.el-select-dropdown__item').at(4).trigger('click')
    await Vue.nextTick()
    await new Promise((resolve) => setTimeout(resolve, 500))

    wrapper.findAll('.dataset-cell').at(1).trigger('click')
    wrapper.findAll('.el-select-dropdown__item').at(8).trigger('click')
    await Vue.nextTick()
    await new Promise((resolve) => setTimeout(resolve, 500))

    wrapper.find('.el-button--primary').trigger('click')
    await Vue.nextTick()

    expect(wrapper.find('.text-danger').exists()).toBe(false)
    // expect(wrapper.vm.$route.name).toBe('datasets-comparison')
    // expect(wrapper.vm.$route.query.ds).toBe(graphqlReturnData.map((ds) => ds.id).sort().join(','))
  })
})
