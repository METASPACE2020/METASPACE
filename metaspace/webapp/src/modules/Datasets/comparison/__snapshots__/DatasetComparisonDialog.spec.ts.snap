// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetComparisonDialog it should match snapshot 1`] = `
<transition-stub name="dialog-fade" class="dataset-comparison-dialog sm-content-page el-dialog-lean" selecteddatasetids="2021-04-06_08h35m04s" style="z-index: 2001;">
  <div class="el-dialog__wrapper">
    <div role="dialog" aria-modal="true" aria-label="dialog" class="el-dialog" style="margin-top: 15vh;">
      <div class="el-dialog__header"><span class="el-dialog__title"></span><button type="button" aria-label="Close" class="el-dialog__headerbtn"><i class="el-dialog__close el-icon el-icon-close"></i></button></div>
      <div class="el-dialog__body">
        <h1>Datasets Comparison</h1>
        <ol class="sm-workflow leading-6 m-0 p-0">
          <li class="sm-workflow-step flex flex-col relative text-gray-600 max-w-measure-3 ml-8 pl-12 border-solid border-0 border-l-2 border-gray-300 transition-colors ease-in-out duration-300 active">
            <p class="sm-workflow-header">Select the datasets</p>
            <form>
              <div class="el-select w-full ">
                <div class="el-select__tags" style="max-width: -32px; width: 100%;">
                  <!---->
                  <transition-group-stub><span class="el-tag el-tag--info el-tag--small el-tag--light"><span class="el-select__tags-text">Untreated_3_434_super_lite_19_31 (1)</span><i class="el-tag__close el-icon-close"></i></span></transition-group-stub>
                  <!---->
                </div>
                <div class="el-input el-input--suffix">
                  <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="" class="el-input__inner" style="height: 40px;">
                  <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----><!----><!----><!----></span>
                  <!----></span>
                  <!---->
                  <!---->
                </div>
                <transition-stub name="el-zoom-in-top">
                  <div class="el-select-dropdown el-popper is-multiple" style="display: none;">
                    <div class="el-scrollbar" style="">
                      <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
                        <ul class="el-scrollbar__view el-select-dropdown__list">
                          <!---->
                          <li class="el-select-dropdown__item"><span>New 3 (1)</span></li>
                          <li class="el-select-dropdown__item"><span>Untreated_3_434_super_lite_19_31 (1)</span></li>
                          <li class="el-select-dropdown__item selected"><span>Untreated_3_434_super_lite_19_31 (1)</span></li>
                        </ul>
                      </div>
                      <div class="el-scrollbar__bar is-horizontal">
                        <div class="el-scrollbar__thumb" style="transform: translateX(0%); webkit-transform: translateX(0%);"></div>
                      </div>
                      <div class="el-scrollbar__bar is-vertical">
                        <div class="el-scrollbar__thumb" style="transform: translateY(0%); webkit-transform: translateY(0%);"></div>
                      </div>
                    </div>
                    <!---->
                  </div>
                </transition-stub>
              </div><button type="button" class="el-button el-button--primary">
                <!---->
                <!----><span>Next</span></button>
            </form>
          </li>
          <li class="sm-workflow-step flex flex-col relative text-gray-600 max-w-measure-3 ml-8 pl-12 border-solid border-0 border-l-2 border-gray-300 transition-colors ease-in-out duration-300">
            <p class="sm-workflow-header">Set the grid arrangement</p>
          </li>
          <li class="sm-workflow-step flex flex-col relative text-gray-600 max-w-measure-3 ml-8 pl-12 border-solid border-0 border-l-2 border-gray-300 transition-colors ease-in-out duration-300">
            <p class="sm-workflow-header">Choose the datasets disposition</p>
          </li>
        </ol>
      </div>
      <!---->
    </div>
  </div>
</transition-stub>
`;
